Exception when running SQL views on DB 17.3 LTS

With relase/9.0.0.

Just tested this on Databricks 17.3 LTS, got this error:

Py4JJavaError: An error occurred while calling o521.execute.
: java.lang.NoClassDefFoundError: org/apache/spark/sql/catalyst/expressions/FoldableUnevaluable
	at au.csiro.pathling.encoders.ValueFunctions.ifArray(ValueFunctions.java:75)
	at au.csiro.pathling.fhirpath.column.DefaultRepresentation.vectorize(DefaultRepresentation.java:109)
	at au.csiro.pathling.fhirpath.column.DefaultRepresentation.vectorize(DefaultRepresentation.java:43)
	at au.csiro.pathling.fhirpath.column.ColumnRepresentation.removeNulls(ColumnRepresentation.java:326)
	at au.csiro.pathling.fhirpath.column.DefaultRepresentation.traverse(DefaultRepresentation.java:121)
	at au.csiro.pathling.fhirpath.collection.ResourceCollection.<init>(ResourceCollection.java:65)
	at au.csiro.pathling.fhirpath.collection.ResourceCollection.build(ResourceCollection.java:118)
	at au.csiro.pathling.fhirpath.execution.BaseResourceResolver.createResource(BaseResourceResolver.java:119)
	at au.csiro.pathling.fhirpath.execution.BaseResourceResolver.resolveSubjectResource(BaseResourceResolver.java:104)
	at au.csiro.pathling.fhirpath.execution.FhirPathEvaluator.createDefaultInputContext(FhirPathEvaluator.java:155)
	at au.csiro.pathling.projection.ProjectionContext.of(ProjectionContext.java:101)
	at au.csiro.pathling.projection.Projection.execute(Projection.java:127)
	at au.csiro.pathling.views.FhirViewExecutor.buildQuery(FhirViewExecutor.java:105)
	at au.csiro.pathling.library.query.DefaultQueryDispatcher.dispatch(DefaultQueryDispatcher.java:40)
	at au.csiro.pathling.library.query.FhirViewQuery.execute(FhirViewQuery.java:123)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:75)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:52)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:397)
	at py4j.Gateway.invoke(Gateway.java:306)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:197)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:117)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassNotFoundException: org.apache.spark.sql.catalyst.expressions.FoldableUnevaluable
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:593)
	at com.databricks.backend.daemon.driver.ClassLoaders$LibraryClassLoader.loadClass(ClassLoaders.scala:152)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 27 more
File <command-315385300460097>, line 1
----> 1 patients = data.view(
      2     "Patient",
      3     select=[
      4         {
      5             "column": [
      6                 {
      7                     "description": "Patient ID",
      8                     "path": "getResourceKey()",
      9                     "name": "id",
     10                 },
     11                 {
     12                     "description": "Birth date",
     13                     "path": "birthDate",
     14                     "name": "birth_date",
     15                 },
     16                 {
     17                     "description": "Postal code",
     18                     "path": "address.postalCode",
     19                     "name": "postal_code",
     20                     "collection": True
     21                 },
     22                 {
     23                     "description": "Deceased time",
     24                     "path": "deceased.ofType(dateTime)",
     25                     "name": "deceased",
     26                 },
     27             ]
     28         },
     29     ],
     30 )
File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.12/site-packages/pathling/datasource.py:105, in DataSource.view(self, resource, select, constants, where, json)
    103 jquery = self._jds.view(resource)
    104 jquery.json(query_json)
--> 105 return self._wrap_df(jquery.execute())

This is the notebook: https://dbc-09a6b098-2f38.cloud.databricks.com/editor/notebooks/315385300460082?o=315729972373157

--- a/encoders/src/main/scala/au/csiro/pathling/encoders/Expressions.scala
+++ b/encoders/src/main/scala/au/csiro/pathling/encoders/Expressions.scala
@@ -29,11 +29,12 @@ import org.apache.spark.sql.Column
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.{TypeCheckFailure, TypeCheckSuccess}
 import org.apache.spark.sql.catalyst.analysis.{TypeCheckResult, UnresolvedException}
-import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.{FoldableUnevaluable, _}
 import org.apache.spark.sql.catalyst.expressions.codegen.Block._
 import org.apache.spark.sql.catalyst.expressions.codegen.{CodeGenerator, CodegenContext, ExprCode, FalseLiteral}
 import org.apache.spark.sql.catalyst.trees.TreePattern.{ARRAYS_ZIP, TreePattern}
 import org.apache.spark.sql.catalyst.util.{ArrayData, GenericArrayData}
+import org.apache.spark.sql.errors.QueryExecutionErrors
 import org.apache.spark.sql.types._
 
 import scala.language.existentials
@@ -282,9 +283,32 @@ case class AttachExtensions(targetObject: Expression,
 }
 
 
+/**
+ * This a mirror of the Spark `Unevaluable`, which is 
+ * needed as a workaround to allow the expression like UnresolvedIfArray to work 
+ * in both Spark 4.0.1 and Spark 4.1.0-rc1 (where FoldableUnevaluable has been removed) 
+ * and which is deployed in Databrics Runtime 17.3 LTS.
+ * <p>
+ * An expression that cannot be evaluated and is not foldable. These expressions 
+ * on't live past analysis or optimization time (e.g. Star)
+ * and should not be evaluated during query planning and execution.
+ */
+trait UnevaluableCopy extends Expression {
+
+  final override def foldable: Boolean = false
+
+  override def eval(input: InternalRow = null): Any =
+    throw SparkException.internalError(s"Cannot evaluate expression: $this")
+
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode =
+    throw SparkException.internalError(s"Cannot generate code for expression: $this")
+
+}
+
 case class UnresolvedIfArray(value: Expression, arrayExpressions: Expression => Expression,
                              elseExpression: Expression => Expression)
-  extends Expression with Unevaluable with NonSQLExpression {
+  extends Expression with UnevaluableCopy with NonSQLExpression {
 
   override def mapChildren(f: Expression => Expression): Expression = {
 
@@ -318,7 +342,7 @@ case class UnresolvedIfArray(value: Expression, arrayExpressions: Expression => 
 
 case class UnresolvedIfArray2(value: Expression, arrayExpressions: Expression => Expression,
                               elseExpression: Expression => Expression)
-  extends Expression with Unevaluable with NonSQLExpression {
+  extends Expression with UnevaluableCopy with NonSQLExpression {
 
   override def mapChildren(f: Expression => Expression): Expression = {
 
@@ -357,7 +381,7 @@ case class UnresolvedIfArray2(value: Expression, arrayExpressions: Expression =>
 
 
 case class UnresolvedUnnest(value: Expression)
-  extends Expression with Unevaluable with NonSQLExpression {
+  extends Expression with UnevaluableCopy with NonSQLExpression {
 
   override def mapChildren(f: Expression => Expression): Expression = {
 
@@ -599,10 +623,11 @@ case class StructProduct(children: Seq[Expression], outer: Boolean = false)
    * @param productSize The total number of product combinations to generate
    * @return An array of InternalRow containing all product combinations
    */
-  private def buildProductArray(inputArrays: Seq[ArrayData], productSize: Int): Array[InternalRow] = {
+  private def buildProductArray(inputArrays: Seq[ArrayData],
+                                productSize: Int): Array[InternalRow] = {
     val result = new Array[InternalRow](productSize)
     val zippedArrays: Seq[(ArrayData, Int)] = inputArrays.zipWithIndex
-    
+
     for (i <- 0 until productSize) {
       val productIndexes = calculateProductIndexes(i, inputArrays)
       val currentRowData = buildRowData(zippedArrays, productIndexes)
@@ -615,13 +640,14 @@ case class StructProduct(children: Seq[Expression], outer: Boolean = false)
    * Calculates the array indexes for a specific product combination.
    *
    * @param productIndex The index of the product combination to calculate
-   * @param inputArrays The sequence of input arrays to index into
+   * @param inputArrays  The sequence of input arrays to index into
    * @return An array of indexes, one for each input array
    */
-  private def calculateProductIndexes(productIndex: Int, inputArrays: Seq[ArrayData]): Array[Int] = {
+  private def calculateProductIndexes(productIndex: Int,
+                                      inputArrays: Seq[ArrayData]): Array[Int] = {
     val productIndexes: Array[Int] = new Array[Int](children.length)
     var productBase = productIndex
-    
+
     for (childIndex <- children.indices) {
       val childArity = inputArrays(childIndex).numElements()
       productIndexes(childIndex) = productBase % childArity
@@ -633,13 +659,14 @@ case class StructProduct(children: Seq[Expression], outer: Boolean = false)
   /**
    * Builds the row data for a single product combination by merging struct fields.
    *
-   * @param zippedArrays The input arrays paired with their indexes
+   * @param zippedArrays   The input arrays paired with their indexes
    * @param productIndexes The array indexes for this product combination
    * @return An array containing the merged field values for the output row
    */
-  private def buildRowData(zippedArrays: Seq[(ArrayData, Int)], productIndexes: Array[Int]): Array[Any] = {
+  private def buildRowData(zippedArrays: Seq[(ArrayData, Int)],
+                           productIndexes: Array[Int]): Array[Any] = {
     val currentRowData = new Array[Any](sizeOfOutput)
-    
+
     zippedArrays.foreach { case (arr, index) =>
       if (!arr.isNullAt(productIndexes(index))) {
         copyStructFields(arr, index, productIndexes(index), currentRowData)
@@ -651,14 +678,15 @@ case class StructProduct(children: Seq[Expression], outer: Boolean = false)
   /**
    * Copies struct fields from an array element to the output row data.
    *
-   * @param arr The source array containing struct elements
-   * @param arrayIndex The index of the source array in the children sequence
-   * @param elementIndex The index of the element within the source array
+   * @param arr            The source array containing struct elements
+   * @param arrayIndex     The index of the source array in the children sequence
+   * @param elementIndex   The index of the element within the source array
    * @param currentRowData The output row data array to copy fields into
    */
-  private def copyStructFields(arr: ArrayData, arrayIndex: Int, elementIndex: Int, currentRowData: Array[Any]): Unit = {
+  private def copyStructFields(arr: ArrayData, arrayIndex: Int, elementIndex: Int,
+                               currentRowData: Array[Any]): Unit = {
     val structData = arr.get(elementIndex, arrayElementTypes(arrayIndex)).asInstanceOf[InternalRow]
-    
+
     for (fi <- 0 until structData.numFields) {
       if (!structData.isNullAt(fi)) {
         currentRowData(offsetsScala(arrayIndex) + fi) = structData.get(fi, null)

